onnxruntime/core/session/inference_session.cc

```diff
@@ -2606,12 +2606,30 @@ common::Status InferenceSession::CheckShapes(const std::string& input_output_nam
                            " Please fix either the inputs/outputs or the model.");
   }
 
+#ifdef USE_QNN
+  // Helper function to check whether QNN EP is used & all nodes are assigned to QNN EP,
+  // and relax the constraint to support batch multiplier on the first dimension.
+  // We will check whether only the Htp backend is used inside QnnModel::ExecuteGraph.
+  auto is_valid_qnn_batch_multiplier = [this](int64_t input_dim, int64_t expected_dim, const Graph& graph) -> bool {
+    if (!AreAllNodesInMainGraphAssignedToOneEp(graph, kQnnExecutionProvider)) {
+      LOGS_IF(input_dim != expected_dim, *session_logger_, WARNING) << "input batch size and expected batch size are different. "
+                                                                    << "Batch multiplier is only supported on the QNN EP, "
+                                                                    << "but some nodes in the graph are assigned to other EPs";
+      return false;
+    }
+    return expected_dim > 0 && input_dim % expected_dim == 0;
+  };
+#endif
+
   InlinedVector<size_t> invalid_dim_indices;
   for (size_t i = 0; i < shape_size; ++i) {
     if (expected_shape[i] < 0) {
       continue;  // this represents a symbolic shape dimension
-    }
-    if (input_output_shape[i] != expected_shape[i]) {
+#ifdef USE_QNN
+    } else if (i == 0 && is_valid_qnn_batch_multiplier(input_output_shape[i], expected_shape[i], model_->MainGraph())) {
+      continue;  // Qnn API supports batch multiplier, but the running batch size must be divisible by the original batch size.
```

Comment:
Can you explain the logic more. Naively if I have a shape of {N, C, H, W} with N being the batch and say {4, 3, 128, 128}. If the batch multiplier is 4 what does the shape change to? 

It feels off to be providing an input shape that does not match what the ONNX model has defined as the required input shape, and to have EP specific logic in ORT code (especially with the move to plugin EPs where the ORT implementation should be EP agnostic). Why can't the QNN EP do that shape re-interpretation internally?